### PR TITLE
fix(logger): add new macro `?LOG_SENSITIVE` and use it to replace some `?LOG`  for security reason

### PR DESCRIPTION
**Pull request by lafirest**
some arguments  passed to external API may contain sensitive data, when the API execution fails, sensitive data may be returned as a part of the failure reason, if this reason is printed to the log, it will lead to sensitive data leakage, so we should check carefully and scan these failed returns

### CODE REVIEW

File: src/emqx_misc.erl [hunk 1]
```diff
@@ -466,6 +468,63 @@ maybe_mute_rpc_log(Node) ->
...
+                     redact(K, V)
+             end, M);
+redact(T) when is_tuple(T) ->
+    Redact = case is_nested_tuple(T) of
+                 {true, Elements} ->
+                     redact(Elements);
+                 {false, [Head | Tail]} ->
+                     [Head | redact(Head, Tail)]
+             end,
+    erlang:list_to_tuple(Redact);
```
Comment (zmstone) [inline src/emqx_misc.erl:501]:
```suggestion
redact(T) when is_tuple(T) ->
    list_to_tuple(redact(tuple_to_list(T)).
```

File: src/emqx_misc.erl [hunk 2]
```diff
@@ -466,6 +468,63 @@ maybe_mute_rpc_log(Node) ->
...
+
+-define(REDACT_VAL, "*****").
+redact_v(V) when is_binary(V) -> <<?REDACT_VAL>>;
+redact_v(V) when is_list(V) ->
+    case io_lib:printable_list(V) of
+        true ->
+            ?REDACT_VAL;
+        _ ->
+            lists:map(fun redact_v/1, V)
+    end;
```
Comment (zmstone) [inline src/emqx_misc.erl:521]: I think it's fare to simply redact the whole value.
unless you have an counter-example ?

File: src/emqx_misc.erl [hunk 3]
```diff
@@ -498,4 +557,44 @@ is_sane_id_test() ->
...
 
+is_sensitive_key_test() ->
+    All = [token, password, secret, passcode, passphrase],
+    lists:foreach(fun(K) ->
+                          ?assert(is_sensitive_key(K)),
+                          ?assert(is_sensitive_key(atom_to_list(K))),
+                          ?assert(is_sensitive_key(atom_to_binary(K)))
+                  end, All).
+
+redact_test() ->
```
Comment (zmstone) [inline src/emqx_misc.erl:568]: Use case generators like belwo:
```
redact_test_() ->
    Case = fun(Key) -> ... end,
    [{case_name(Type, Key), fun() -> Case(Key) end} || Key <- Keys, Type <- Types].

case_name(Type, Key) ->
    atom_to_list(Type) ++ "-" ++ key_to_list(Key).
```

File: src/emqx_misc.erl (hunk 1 above)
Comment (lafirest) [reply src/emqx_misc.erl:501]: I think we still need to iterate over the tuple,
e.g.
```
{{shutdown,{failed_to_start_child,worker_sup,
                                  {shutdown,{failed_to_start_child,{worker,1},
                                                                   {connection_error,econnrefused}}}}},
 {child,undefined,
        {pool_sup,'backend_redis:resource:723930'},
        {ecpool_pool_sup,start_link,
                         ['backend_redis:resource:723930',emqx_backend_redis_actions,
                          [{sentinel,[]},
                           {pool_size,1},
                           {database,0},
                           {password,"12312321321312321"},
                           {auto_reconnect,2},
                           {host,"127.0.0.1"},
                           {port,6379},
                           {options,[]}]]},
        transient,false,infinity,supervisor,
        [ecpool_pool_sup]}}
```

File: src/emqx_misc.erl (hunk 2 above)
Comment (lafirest) [reply src/emqx_misc.erl:521]: This is used to solve that can't distinguish list/string when traversing tuple

File: src/emqx_misc.erl [hunk 4]
```diff
@@ -498,4 +549,46 @@ is_sane_id_test() ->
...
```
Comment (zmstone) [inline src/emqx_misc.erl:560]:
```suggestion
redact_test_() ->
```

File: src/emqx_misc.erl [hunk 5]
```diff
@@ -498,4 +549,46 @@ is_sane_id_test() ->
...
+    %% 3 level nested
+    ?assertEqual([#{opts => [{Key, ?REDACT_VAL}]}], redact([#{opts => [{Key, foo}]}])),
+    ?assertEqual([{opts, [{Key, ?REDACT_VAL}]}], redact([{opts, [{Key, foo}]}])),
+    ?assertEqual([{opts, [#{Key => ?REDACT_VAL}]}], redact([{opts, [#{Key => foo}]}]))
+    end,
+
+    Types = [atom, string, binary],
+    Keys = [token, password, secret, passcode, passphrase],
+    [{case_name(Type, Key), fun() -> Case(Type, Key) end} || Key <- Keys, Type <- Types],
+    ok.
```
Comment (zmstone) [inline src/emqx_misc.erl:589]:
```suggestion
    [{case_name(Type, Key), fun() -> Case(Type, Key) end} || Key <- Keys, Type <- Types].
```

**Review by zmstone — APPROVED** [review]

**Review by zmstone — CHANGES_REQUESTED** [review]
update changes/v4.3.22-en|zh.md

File: changes/v4.3.22-en.md [hunk 6]
```diff
@@ -2,6 +2,9 @@
 
 - Add a warning log if the ACL check failed for subscription [#9124](https://github.com/emqx/emqx/pull/9124).
   This is to make the ACL deny logging for subscription behave the same as for publish.
+- Enhanced security of logs containing sensitive data. [#9189](https://github.com/emqx/emqx/pull/9189).
+
+  If the log contains sensitive data such as passwords, the data will be obfuscated
```
Comment (zmstone) [inline changes/v4.3.22-en.md:7]:
```suggestion
- Enhanced security of logs containing sensitive data [#9189](https://github.com/emqx/emqx/pull/9189).
  If the log contains sensitive data such as passwords, the data will be obfuscated
```

File: changes/v4.3.22-zh.md [hunk 7]
```diff
@@ -2,6 +2,9 @@
 
 - 订阅时，如果 ACL 检查不通过，打印一个警告日志 [#9124](https://github.com/emqx/emqx/pull/9124)。
   该行为的改变主要是为了跟发布失败时的行为保持一致。
+- 增强包含敏感数据的日志的安全性 [#9189](https://github.com/emqx/emqx/pull/9189).
+
+  如果日志中包含密码等敏感数据,将会对这些数据进行模糊化处理
```
Comment (zmstone) [inline changes/v4.3.22-zh.md:7]:
```suggestion

- 增强包含敏感数据的日志的安全性 [#9189](https://github.com/emqx/emqx/pull/9189)。
  如果日志中包含敏感关键词，例如 `password`，那么关联的数据回被模糊化处理，替换成 `******`。
```

**Review by zmstone — COMMENTED** [review]
looks good, but I should not approve it as I contributed the initial draft.

**Review by zmstone — APPROVED** [review]

**Review by thalesmg — APPROVED** [review]

File: changes/v4.3.22-zh.md [hunk 8]
```diff
@@ -5,6 +5,9 @@
 
 - 基于 JWT 的 ACL 支持 `all` 动作，指定同时适用于 `pub` 和 `sub` 两个动作的规则列表 [#9044](https://github.com/emqx/emqx/pull/9044)。
 
+- 增强包含敏感数据的日志的安全性 [#9189](https://github.com/emqx/emqx/pull/9189).
```
Comment (terry-xiaoyu) [inline changes/v4.3.22-zh.md:8]:
```suggestion
- 增强包含敏感数据的日志的安全性 [#9189](https://github.com/emqx/emqx/pull/9189)。
```

File: changes/v4.3.22-en.md [hunk 9]
```diff
@@ -5,6 +5,9 @@
 
 - JWT ACL claim supports `all` action to imply the rules applie to both `pub` and `sub` [#9044](https://github.com/emqx/emqx/pull/9044).
 
+- Added a log censor to avoid logging sensitive data [#9189](https://github.com/emqx/emqx/pull/9189).
+  If the data to be logged is a map or key-value list which contains sensitive key words such as `password`, the value is obfuscated as `******`
```
Comment (terry-xiaoyu) [inline changes/v4.3.22-en.md:9]:
```suggestion
  If the data to be logged is a map or key-value list which contains sensitive key words such as `password`, the value is obfuscated as `******`.
```

**Review by terry-xiaoyu — DISMISSED** [review]

File: src/emqx_misc.erl [hunk 10]
```diff
@@ -466,6 +468,67 @@ maybe_mute_rpc_log(Node) ->
...
+is_sensitive_key(aws_secret_access_key) -> true;
+is_sensitive_key("aws_secret_access_key") -> true;
+is_sensitive_key(<<"aws_secret_access_key">>) -> true;
+is_sensitive_key(secret_key) -> true;
+is_sensitive_key("secret_key") -> true;
+is_sensitive_key(<<"secret_key">>) -> true;
+is_sensitive_key(bind_password) -> true;
+is_sensitive_key("bind_password") -> true;
+is_sensitive_key(<<"bind_password">>) -> true;
+is_sensitive_key(_) -> false.
```
[the remark this answers is not in the document]
Comment (zmstone) [reply src/emqx_misc.erl:498]: sasl allows wrapped password: https://github.com/emqx/emqx-enterprise/pull/1454

File: src/emqx_misc.erl [hunk 11]
```diff
@@ -498,4 +561,41 @@ is_sane_id_test() ->
...
+              string -> erlang:atom_to_list(KeyT);
+              binary -> erlang:atom_to_binary(KeyT)
+          end,
+
+    ?assert(is_sensitive_key(Key)),
+
+    %% direct
+    ?assertEqual({Key, ?REDACT_VAL}, redact({Key, foo})),
+    ?assertEqual(#{Key => ?REDACT_VAL}, redact(#{Key => foo})),
+
```
Comment (zmstone) [inline src/emqx_misc.erl:577]: add:
```suggestion
? assertEqual({[{Key, ?REDACT_VAL}], bar}, redact({[{Key, foo}], bar})).
```

**Review by zmstone — APPROVED** [review]

**Pull request merged**